### PR TITLE
create `ExpectAtLeastTwoWorkerNodesWithCPUManager` as check and deduce number of nodes from usage inside `repeated_test.sh`

### PR DESCRIPTION
**Pull request by dhiller**
**What this PR does / why we need it**:
Extracts `ExpectAtLeastTwoWorkerNodesWithCPUManager` as func and puts it into checks package. Replaces explicit node fetches with using the func where applicable.

Adds func name to grep call in `repeated_test.sh` to make sure kubevirtci instance has two cpumanager nodes if tests that require them are run.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
/cc @xpivarc @jean-edouard 

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by dhiller** [comment]
> Generally speaking: how about adding a label decorator for it? Thanks

Makes sense I think. But how would we change detecting whether tests that have it are being selected? Would we just grep for the selector instead? Or would we dry-run with selector and see whether there are still tests inside the set? (I think that the latter would add quite a bit of complexity)

**Comment by fossedihelm** [comment]
> > Generally speaking: how about adding a label decorator for it? Thanks
> 
> Makes sense I think. But how would we change detecting whether tests that have it are being selected? Would we just grep for the selector instead? Or would we dry-run with selector and see whether there are still tests inside the set? (I think that the latter would add quite a bit of complexity)

I think we can directly grep for the label decorator! This will not change the current behavior. Don't you think?

**Comment by dhiller** [comment]
> > Generally speaking: how about adding a label decorator for it? Thanks
> 
> Makes sense I think. But how would we change detecting whether tests that have it are being selected? Would we just grep for the selector instead? Or would we dry-run with selector and see whether there are still tests inside the set? (I think that the latter would add quite a bit of complexity)

@fossedihelm I am mentioning this that `repeated_test.sh` has a file centric selection approach in contrast to the "normal" filtering approach via labels and test names.

**Comment by dhiller** [comment]
> > > Generally speaking: how about adding a label decorator for it? Thanks
> > 
> > 
> > Makes sense I think. But how would we change detecting whether tests that have it are being selected? Would we just grep for the selector instead? Or would we dry-run with selector and see whether there are still tests inside the set? (I think that the latter would add quite a bit of complexity)
> 
> I think we can directly grep for the label decorator! This will not change the current behavior. Don't you think?

No, I agree. I was just asking for a bit more detail, which you have delivered with this answer. Thanks for that!

**Comment by dhiller** [comment]
@fossedihelm I've updated the PR with your suggestion, PTAL! Thank you!

File: tests/framework/checks/expects.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+func GetWorkerNodesWithCPUManagerEnabled(virtClient kubecli.KubevirtClient) []v1.Node {
+	ginkgo.By("getting the list of worker nodes that have cpumanager enabled")
+	nodeList, err := virtClient.CoreV1().Nodes().List(context.TODO(), v12.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=,%s=%s", workerLabel, "cpumanager", "true"),
+	})
+	gomega.Expect(err).ToNot(gomega.HaveOccurred())
+	gomega.Expect(nodeList).ToNot(gomega.BeNil())
+	return nodeList.Items
+}
```
Comment (fossedihelm) [inline tests/framework/checks/expects.go:61]: I'm a little disturbed that we're following two different paths to achieve the same goal.
ref. https://github.com/kubevirt/kubevirt/blob/d7371feb56fcb138806b1b6e5b6ab154e900a2e0/tests/framework/checks/skips.go#L52-L75

But this is out of scope of this PR maybe!
I am thinking about a refactor about these

File: tests/framework/checks/expects.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+// ExpectAtLeastTwoNodes uses gomega.Expect to verify that the given node list contains at least two elements. The
+// optionalDescription can be used to override the default description that notifies the reader of "not enough worker
+// nodes with cpumanager"
+func ExpectAtLeastTwoNodes(nodes []v1.Node, optionalDescription string) bool {
+	if optionalDescription == "" {
+		optionalDescription = descriptionTwoWorkerNodesWCPUManagerRequired
+	}
+	ginkgo.By("ensuring at least 2 worker nodes have cpumanager")
+	return gomega.Expect(len(nodes)).To(gomega.BeNumerically(">=", 2), optionalDescription)
+}
```
Comment (fossedihelm) [inline tests/framework/checks/expects.go:51]: IMHO the file should not give preference to any specific check. It should be agnostic.
The optionalDescription in currently never provided, and also, the `ginkgo.By` sentence ignores it.
This is what I suggest for this piece of code:
- make the description mandatory since the function just says `ExpectAtLeastTwoNodes`

Let me know what do you think. Thanks
```suggestion
const (
	workerLabel                                  = "node-role.kubernetes.io/worker"
	DescriptionTwoWorkerNodesWCPUManagerRequired = "at least two worker nodes with cpumanager are required for migration"
)

func ExpectAtLeastTwoWorkerNodesWithCPUManager(virtClient kubecli.KubevirtClient) {
	ExpectAtLeastTwoNodes(GetWorkerNodesWithCPUManagerEnabled(virtClient), DescriptionTwoWorkerNodesWCPUManagerRequired)
}

// ExpectAtLeastTwoNodes uses gomega.Expect to verify that the given node list contains at least two elements.
func ExpectAtLeastTwoNodes(nodes []v1.Node, description string) bool {
	ginkgo.By("ensuring at least 2 nodes")
	return gomega.Expect(len(nodes)).To(gomega.BeNumerically(">=", 2), description)
}
```

**Review by fossedihelm — COMMENTED** [review]
@dhiller Looks good overall, just some suggestions below.
Also, can you squash the last 2 commits in a single one? Thank you!

File: tests/framework/checks/expects.go (hunk 1 above)
Comment (dhiller) [reply tests/framework/checks/expects.go:61]: Actually I don't really understand why at first we are asking whether the feature gate is there and then again ask for cpumanager being present, the former sounds superfluous to me. But yes, this is out of scope to me also.

File: tests/framework/checks/expects.go (hunk 2 above)
Comment (dhiller) [reply tests/framework/checks/expects.go:51]: Great suggestion, makes sense, thanks!

**Comment by dhiller** [comment]
@fossedihelm I've addressed your comments, PTAL, thank you!

File: tests/framework/checks/expects.go (hunk 1 above)
Comment (fossedihelm) [reply tests/framework/checks/expects.go:57]: Well, actually it has sense to me! Even if the CPUManager FG is enabled, a node could not have the cpu manager running.

**Review by fossedihelm — APPROVED** [review]
Thanks @dhiller

File: tests/framework/checks/expects.go (hunk 1 above)
Comment (dhiller) [reply tests/framework/checks/expects.go:57]: Yes, that's why I think the FG check is superfluous, no?

Comment (fossedihelm) [reply tests/framework/checks/expects.go:57]: Yeah! My bad! Misread xD

**Comment by dhiller** [comment]
@fossedihelm I've rebased the PR, PTAL! Thank you!

**Comment by fossedihelm** [comment]
/lgtm
Sorry, I missed it

**Comment by dhiller** [comment]
@jean-edouard @xpivarc could you take a look, please? Thank you!

File: tests/framework/checks/expects.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+func ExpectAtLeastTwoWorkerNodesWithCPUManager(virtClient kubecli.KubevirtClient) {
+	ExpectAtLeastTwoNodes(GetWorkerNodesWithCPUManagerEnabled(virtClient), DescriptionTwoWorkerNodesWCPUManagerRequired)
+}
+
+// ExpectAtLeastTwoNodes uses gomega.Expect to verify that the given node list contains at least two elements. The
+// description is added to the default description via ginkgo.By
+func ExpectAtLeastTwoNodes(nodes []v1.Node, description string) bool {
+	ginkgo.By(fmt.Sprintf("expect at least 2 nodes - %s", description))
+	return gomega.Expect(len(nodes)).To(gomega.BeNumerically(">=", 2), description)
```
Comment (xpivarc) [inline tests/framework/checks/expects.go:46]: Won't this fail?

File: tests/migration/migration.go [hunk 4]
```diff
@@ -3142,15 +3141,8 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
...
-			By("getting the list of worker nodes that have cpumanager enabled")
-			nodeList, err := virtClient.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{
-				LabelSelector: fmt.Sprintf("%s=,%s=%s", workerLabel, "cpumanager", "true"),
-			})
-			Expect(err).ToNot(HaveOccurred())
-			Expect(nodeList).ToNot(BeNil())
-			nodes = nodeList.Items
-			Expect(len(nodes)).To(BeNumerically(">=", 2), "at least two worker nodes with cpumanager are required for migration")
+			nodes = checks.GetWorkerNodesWithCPUManagerEnabled(virtClient)
+			checks.ExpectAtLeastTwoNodes(nodes, checks.DescriptionTwoWorkerNodesWCPUManagerRequired)
```
Comment (xpivarc) [inline tests/migration/migration.go:3145]: Isn't this just `ExpectAtLeastTwoWorkerNodesWithCPUManager`? Also above we use `checks.SkipTestIfNotEnoughNodesWithCPUManager(2)`

File: tests/framework/checks/expects.go (hunk 3 above)
Comment (dhiller) [reply tests/framework/checks/expects.go:46]: Oh, now I see :grin:

File: tests/framework/checks/expects.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	DescriptionTwoWorkerNodesWCPUManagerRequired = "at least two worker nodes with cpumanager are required for migration"
+)
+
+func ExpectAtLeastTwoWorkerNodesWithCPUManager(virtClient kubecli.KubevirtClient) {
+	ExpectAtLeastTwoNodes(GetWorkerNodesWithCPUManagerEnabled(virtClient), DescriptionTwoWorkerNodesWCPUManagerRequired)
+}
+
+// ExpectAtLeastTwoNodes uses gomega.Expect to verify that the given node list contains at least two elements. The
+// description is added to the default description via ginkgo.By
+func ExpectAtLeastTwoNodes(nodes []v1.Node, description string) bool {
```
Comment (xpivarc) [inline tests/framework/checks/expects.go:44]: This does not seem to be exported anymore.

File: tests/framework/checks/expects.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	ginkgo.By(fmt.Sprintf("expect at least 2 nodes - %s", description))
+	return gomega.Expect(len(nodes)).To(gomega.BeNumerically(">=", 2), description)
+}
+
+func GetWorkerNodesWithCPUManagerEnabled(virtClient kubecli.KubevirtClient) []v1.Node {
```
Comment (xpivarc) [inline tests/framework/checks/expects.go:49]: This would be more fit to `tests/libnode`

File: tests/framework/checks/expects.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+package checks
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+	v1 "k8s.io/api/core/v1"
+	v12 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (xpivarc) [inline tests/framework/checks/expects.go:29]: nit: A more fit name and used commonly in our repo is `metav1`

File: tests/framework/checks/expects.go (hunk 3 above)
Comment (dhiller) [reply tests/framework/checks/expects.go]: Removed the return value

File: tests/framework/checks/expects.go (hunk 5 above)
Comment (dhiller) [reply tests/framework/checks/expects.go]: I've inlined the func.

File: tests/framework/checks/expects.go (hunk 6 above)
Comment (dhiller) [reply tests/framework/checks/expects.go]: Moved

**Comment by dhiller** [comment]
@xpivarc thank you for the review! I've addressed comments, PTAL, thank you!

**Comment by dhiller** [comment]
> @xpivarc thank you for the review! I've addressed comments, PTAL, thank you!

@xpivarc poking you again

**Comment by dhiller** [comment]
@xpivarc I think I have addressed all the comments here, did I miss anything?

**Review by EdDev — COMMENTED** [review]
Sorry for the later review.
If this is too late, you can ignore my comments.
